onnxruntime/test/util/compare_ortvalue.cc

```diff
@@ -284,12 +284,18 @@ bool AreShapesEqual(const std::vector<int64_t>& real_shape, const ::ONNX_NAMESPA
   if (len < 0) return false;
   if (real_shape.size() != static_cast<size_t>(len)) return false;
   for (int i = 0; i != len; ++i) {
-    if (!expected_shape.dim(i).has_dim_value()) {
-      // symbolic shape, cannot validate it right now, assume it matches every thing
-      continue;
+    const auto& dim = expected_shape.dim(i);
+    switch (dim.value_case()) {
+      case ONNX_NAMESPACE::TensorShapeProto::Dimension::kDimValue:
+        if (dim.dim_value() != real_shape[i]) return false;
+        break;
+      case ONNX_NAMESPACE::TensorShapeProto::Dimension::kDimParam:
+        // symbolic shape, cannot validate it right now, assume it matches every thing
+        continue;
+        break;
+      default : assert(false);
```

Comment:
The default should continue;

Reply:
I think the default should be to return false. If the shape has an unknown dimension (no dim_param or dim_value) it can't be considered equal to another shape. 

---
In reply to: [322018374](https://github.com/microsoft/onnxruntime/pull/1778#discussion_r322018374) [](ancestors = 322018374)

Reply:
This function is only used in the onnx_test_runner. Sometimes, the model doesn't have input/output shape info, but when run the model with actual data, the output should have.  In this case, we can't fail the test.

Reply:
Got it. Maybe it would be clearer to not use a switch statement if the logic is 'check if dim_value, ignore otherwise and assume matches'

---
In reply to: [322030188](https://github.com/microsoft/onnxruntime/pull/1778#discussion_r322030188) [](ancestors = 322030188)

Reply:
It is not like it does not have shape bc the dimension is *there*, otherwise it would be UNSET. One can't avoid switch statement bc anything else would either be an equivalent OR it would simply punt on other cases, being either UNSET or being DimParam. Current code simply assumes it is DimParam, but it is wrong. Let's discuss it.

---
In reply to: [322030188](https://github.com/microsoft/onnxruntime/pull/1778#discussion_r322030188) [](ancestors = 322030188)

Reply:
I have a proposal to handle UNSET case as returning true as being OK according the case Changming outilned and keep assert in the default in case we have a new oneof (unlikely but that's what assert is for).

---
In reply to: [322350853](https://github.com/microsoft/onnxruntime/pull/1778#discussion_r322350853) [](ancestors = 322350853,322030188)